taskrun/TaskManager.py

```diff
@@ -235,112 +245,173 @@ def task_killed(self, task):
+
+  def _clear_waiting_and_ready(self):
+    """
+    Clears the waiting and ready tasks lists. Adds the tasks to the filtered
+    list.
+    """
+    # clear out waiting and ready lists
+    self._filter_tasks.extend(self._waiting_tasks)
+    self._waiting_tasks = []
+    for priority in range(self._priority_levels):
+      self._filter_tasks.extend(self._ready_tasks[priority])
+      self._ready_tasks[priority] = []
+
+  def _remove_dependent_tasks(self, task):
+    """
+    Removes the dependent tasks of the specified task. Adds the tasks to the
+    filtered list.
+    """
+    # remove all tasks that depend on this task (BFS)
+    visit = []
+    visit.extend(task.get_dependents())
+    visited = set()
+    while len(visit) > 0:
+      curr = visit.pop()
+      for priority in range(self._priority_levels):
+        assert curr not in self._ready_tasks[priority]
+      assert curr not in self._running_tasks
+      visited.add(curr)
+      for dep in curr.get_dependents():
+        if dep not in visited:
+          visit.append(dep)
+      if curr in self._waiting_tasks:
+        self._waiting_tasks.remove(curr)
+        self._filter_tasks.append(curr)
+
   def _task_done(self, task):
     """
-    This method is called by task_completed and task_failed.
+    This method is called by task_completed and task_error.
     This method removes the task from the running list and gives back all
     resources it consumed.
 
-    WARNING: this method must be called while locked on the condition variable
+    WARNING: this method must be called while locked on the condition variable.
 
     Args:
       task (Task) : the task that finished
     """
     assert self._running is True
 
     # remove task from running lists
-    self._running_tasks.remove(task)
+    if task:
+      self._running_tasks.remove(task)
 
     # give back resources
-    if not task.bypass:
+    if task and not task.bypass:
       if self._resource_manager is not None:
         self._resource_manager.done(task)
 
     # notify waiting thread
     self._condition_variable.notify()
 
+  def _terminate(self):
+    """
+    Executes the procedure for graceful forced shutdown.
+    """
+    self._failure_mode = FailureMode.AGGRESSIVE_FAIL
+    self._task_error(None, None)
+
+  def _set_signal_handlers(self):
+    """
+    Sets the signal handlers for SIGINT and SIGTERM to gracefully shutdown when
+    received. SIGINT also includes a 3 seconds guard window.
+    """
+    handler_block = [False]
+    def handler(signum, frame):
+      if handler_block[0]:
+        return
+      handler_block[0] = True
+      if os.getpid() != self._pid:
+        os._exit(-1)
+        return
+      now = datetime.datetime.now()
+      delta = (now - self._last_signal_time).total_seconds()
+      self._last_signal_time = now
+      if signum == signal.SIGTERM or delta < 3.0:
+        signal.signal(signal.SIGINT, signal.SIG_DFL)
+        signal.signal(signal.SIGTERM, signal.SIG_DFL)
```

Comment:
is this necessary, give that the main thread will call _reset_signal_handlers later and handler_block will block future signals?

Reply:
Yes, because the signal blocker turns on ignoring signals. SIG_DFL will kill the process. With SIG_DFL, it should be impossible to hang taskrun. Without it, a task that ignores signals would hang all of taskrun.

Reply:
Thanks. Call _reset_signal_handlers here instead?

Reply:
Done